arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,223 @@
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <Inspection/VPackWithErrorT.h>
+#include <memory>
+#include <iostream>
+#include <string_view>
+#include <type_traits>
+
+#include "ActorBase.h"
+#include "HandlerBase.h"
+#include "Message.h"
+#include "MPSCQueue.h"
+
+namespace arangodb::pregel::actor {
+
+namespace {
+template<typename Runtime, typename A>
+concept IncludesAllActorRelevantTypes =
+    std::is_class<typename A::State>::value &&
+    std::is_class<typename A::template Handler<Runtime>>::value &&
+    std::is_class<typename A::Message>::value;
+template<typename A>
+concept IncludesName = requires() {
+  { A::typeName() } -> std::convertible_to<std::string_view>;
+};
+template<typename Runtime, typename A>
+concept HandlerInheritsFromBaseHandler =
+    std::is_base_of < HandlerBase<Runtime, typename A::State>,
+typename A::template Handler<Runtime> > ::value;
+template<typename Runtime, typename A>
+concept MessageIsVariant = requires(ActorPID pid,
+                                    std::shared_ptr<Runtime> runtime,
+                                    typename A::State state,
+                                    typename A::Message message) {
+  {std::visit(
+      typename A::template Handler<Runtime>{
+          {pid, pid, std::move(std::make_unique<typename A::State>(state)),
+           runtime}},
+      message)};
+};
+template<typename A>
+concept IsInspectable = requires(typename A::Message message,
+                                 typename A::State state) {
+  {fmt::format("{}", message)};
+  {fmt::format("{}", state)};
+};
+};  // namespace
+template<typename Runtime, typename A>
+concept Actorable = IncludesAllActorRelevantTypes<Runtime, A> &&
+    IncludesName<A> && HandlerInheritsFromBaseHandler<Runtime, A> &&
+    MessageIsVariant<Runtime, A> && IsInspectable<A>;
+
+template<typename Runtime, typename Config>
+requires Actorable<Runtime, Config>
+struct Actor : ActorBase {
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState)
+      : pid(pid), runtime(runtime), state(std::move(initialState)) {}
+
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState,
+        std::size_t batchSize)
+      : pid(pid),
+        runtime(runtime),
+        state(std::move(initialState)),
+        batchSize(batchSize) {}
+
+  ~Actor() = default;
+
+  auto typeName() -> std::string_view override { return Config::typeName(); };
+
+  void process(ActorPID sender,
+               std::unique_ptr<MessagePayloadBase> msg) override {
+    auto* ptr = msg.release();
+    if (auto* m = dynamic_cast<MessagePayload<typename Config::Message>*>(ptr);
```

Comment:
dynamic cast to resolve type of message is a bad idea.
It's very slow, and it's additional place where we require rtti

For example you can add tag virtual member which will be return enum class and then make switch on it.
And do static_cast.
It's more effective and universal.
Also it will be single virtual call on all process, instead of N dynamic_cast

Reply:
I implemented Manuel's suggestion in https://github.com/arangodb/arangodb/pull/17970#discussion_r1073501129

Reply:
My comment about dynamic_cast, in your fix you still make dynamic_cast in worst case N-1 times

Reply:
I'm not sure I understand your idea with the enum: I understand that you want to add a virtual method to MessgePayloadBase, e.g.:
```
virtual auto type() -> Type;
```
where `Type` is an enum. Then this enum has to include ActorError and the Message types of all actors in this runtime. If I understood this correct, there is a usability problem with this:
We would like only the Actor itself to know its own Message types it can handle: that way, when you define your own specific actor - e.g. the TrivialActor (in tests/Pregel/Actor/Actors/TrivialActor.h) - you only define the message types once in the specific actor definition (see TrivialMessages in TrivialActor.h). With the enum you would need to add an alternative to the enum each time you define a new specific actor.